File: python/tests/test_effective_hamiltonian.py
"""Tests for the EffectiveHamiltonianConstructor bindings and registry."""

# --------------------------------------------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE.txt in the project root for license information.
# --------------------------------------------------------------------------------------------

from types import SimpleNamespace

import numpy as np
import pytest

from qdk_chemistry import algorithms
from qdk_chemistry.algorithms import (
    EffectiveHamiltonianConstructor,
    QdkSchriefferWolffPT2Constructor,
)
from qdk_chemistry.constants import ANGSTROM_TO_BOHR
from qdk_chemistry.data import Hamiltonian, Orbitals, Structure
from qdk_chemistry.data.symmetry import SymmetryLabel, axes, spin_index_set

_TYPE = "effective_hamiltonian_constructor"

_WATER = Structure(
    ["O", "H", "H"],
    np.array(
        [
            [0.0, -0.0757918436, 0.0],
            [0.866811829, 0.6014357793, 0.0],
            [-0.866811829, 0.6014357793, 0.0],
        ]
    )
    * ANGSTROM_TO_BOHR,
)


@pytest.fixture(scope="module")
def water_window():
    """Water/STO-3G with a CAS(2e,2o) reference and a Hamiltonian over ``P`` plus one orbital."""
    alpha = SymmetryLabel([axes.alpha()])
    _, wfn_hf = algorithms.create("scf_solver").run(_WATER, 0, 1, "sto-3g")
    orbitals = wfn_hf.get_orbitals()
    norb = orbitals.get_num_molecular_orbitals()

    selector = algorithms.create("active_space_selector", "qdk_valence")
    selector.settings().set("num_active_electrons", 2)
    selector.settings().set("num_active_orbitals", 2)
    reference = selector.run(wfn_hf)

    ref_orbs = reference.get_orbitals()
    kept = [int(i) for i in ref_orbs.active_indices().indices(alpha)]
    core = [int(i) for i in ref_orbs.inactive_indices().indices(alpha)]
    used = set(kept) | set(core)
    window = sorted([*kept, next(i for i in range(norb) if i not in used)])

    window_orbitals = Orbitals(
        orbitals.coefficients().block([alpha, alpha]),
        orbitals.energies().block([alpha]),
        None,
        orbitals.get_basis_set(),
        active_indices=spin_index_set(norb, window, window),
        inactive_indices=spin_index_set(norb, core, core),
    )
    h_window = algorithms.create("hamiltonian_constructor").run(window_orbitals)
    return SimpleNamespace(
        reference=reference,
        hamiltonian=h_window,
        kept_space=spin_index_set(norb, kept, kept),
        window_space=spin_index_set(norb, window, window),
        kept_in_window=[window.index(i) for i in kept],
    )


class TestEffectiveHamiltonianConstructor:
    """Registry, factory, and settings coverage for the downfolding bindings."""

    def test_factory_registration(self):
        """qdk_swpt2 is registered as an effective-Hamiltonian constructor and is the default."""
        available = algorithms.available(_TYPE)
        assert isinstance(available, list)
        assert "qdk_swpt2" in available

        # default and explicit creation both yield the swpt2 constructor
        default = algorithms.create(_TYPE)
        assert isinstance(default, EffectiveHamiltonianConstructor)
        assert default.name() == "qdk_swpt2"
        assert default.type_name() == _TYPE

        explicit = algorithms.create(_TYPE, "qdk_swpt2")
        assert explicit.name() == "qdk_swpt2"

        # a nonexistent name raises
        with pytest.raises((KeyError, RuntimeError)):
            algorithms.create(_TYPE, "nonexistent_downfolder")

    def test_direct_construction(self):
        """The concrete class constructs directly."""
        constructor = QdkSchriefferWolffPT2Constructor()
        assert isinstance(constructor, EffectiveHamiltonianConstructor)
        assert constructor.name() == "qdk_swpt2"

    def test_settings_knobs(self):
        """Denominator/regularization settings expose their defaults and are settable."""
        constructor = algorithms.create(_TYPE, "qdk_swpt2")
        settings = constructor.settings()
        # sigma^2 regularization is on by default at the constructor layer
        assert settings.get("regularizer_sigma2") == pytest.approx(1.0)
        assert settings.get("semicanonicalize") is True
        assert settings.get("fold_above_two_body") is True
        assert settings.get("max_folded_occupation_deviation") == pytest.approx(0.5)

        settings.set("regularizer_sigma2", 0.4)
        assert constructor.settings().get("regularizer_sigma2") == pytest.approx(0.4)

    def test_accepts_mean_field_hf_reference(self):
        """A mean-field HF reference (no active 1-RDM) is accepted directly."""
        alpha = SymmetryLabel([axes.alpha()])
        water = Structure(
            ["O", "H", "H"],
            np.array(
                [
                    [0.0, -0.0757918436, 0.0],
                    [0.866811829, 0.6014357793, 0.0],
                    [-0.866811829, 0.6014357793, 0.0],
                ]
            )
            * ANGSTROM_TO_BOHR,
        )
        _, wfn_hf = algorithms.create("scf_solver").run(water, 0, 1, "sto-3g")
        orbitals = wfn_hf.get_orbitals()

        selector = algorithms.create("active_space_selector", "qdk_valence")
        selector.settings().set("num_active_electrons", 2)
        selector.settings().set("num_active_orbitals", 2)
        reference = selector.run(wfn_hf)  # single-determinant HF, no 1-RDM

        def with_active_space(orb, active, inactive):
            """Return a copy of `orb` with the given active/inactive index sets."""
            n = orb.get_num_molecular_orbitals()
            return Orbitals(
                orb.coefficients().block([alpha, alpha]),
                orb.energies().block([alpha]),
                None,
                orb.get_basis_set(),
                active_indices=spin_index_set(n, list(active), list(active)),
                inactive_indices=spin_index_set(n, list(inactive), list(inactive)),
            )

        # small window = active P + the lowest orbital outside P and the core
        ref_orbs = reference.get_orbitals()
        active = list(ref_orbs.active_indices().indices(alpha))
        core = list(ref_orbs.inactive_indices().indices(alpha))
        norb = orbitals.get_num_molecular_orbitals()
        used = set(active) | set(core)
        window = sorted([*active, next(i for i in range(norb) if i not in used)])

        h_window = algorithms.create("hamiltonian_constructor").run(with_active_space(orbitals, window, core))
        downfolder = algorithms.create(_TYPE, "qdk_swpt2")
        p_space = spin_index_set(norb, active, active)  # kept space P as an index set
        h_eff = downfolder.run(reference, h_window, p_space)  # must not raise
        assert isinstance(h_eff, Hamiltonian)

    def test_full_window_spanning_reference_core(self):
        """Downfold a full-orbital window (the natural ``ham.run(orbitals)`` path).

        The core orbitals are reference core but lie inside ``W``, so the
        downfold folds them itself instead of inheriting them from the window
        Hamiltonian, whose own inactive set is empty here. They must still come
        back labelled inactive, exactly once each.
        """
        alpha = SymmetryLabel([axes.alpha()])
        water = Structure(
            ["O", "H", "H"],
            np.array(
                [
                    [0.0, -0.0757918436, 0.0],
                    [0.866811829, 0.6014357793, 0.0],
                    [-0.866811829, 0.6014357793, 0.0],
                ]
            )
            * ANGSTROM_TO_BOHR,
        )
        _, wfn_hf = algorithms.create("scf_solver").run(water, 0, 1, "sto-3g")

        selector = algorithms.create("active_space_selector", "qdk_valence")
        selector.settings().set("num_active_electrons", 2)
        selector.settings().set("num_active_orbitals", 2)
        reference = selector.run(wfn_hf)

        ref_orbs = reference.get_orbitals()
        active = [int(i) for i in ref_orbs.active_indices().indices(alpha)]
        core = [int(i) for i in ref_orbs.inactive_indices().indices(alpha)]

        # window over the full orbital set (spans the reference core)
        h_full = algorithms.create("hamiltonian_constructor").run(wfn_hf.get_orbitals())
        # P == reference active space: pass its index set straight through
        h_eff = algorithms.create(_TYPE, "qdk_swpt2").run(
            reference, h_full, ref_orbs.active_indices()
        )  # must not raise

        assert isinstance(h_eff, Hamiltonian)
        emitted = h_eff.get_orbitals()
        assert list(emitted.active_indices().indices(alpha)) == active
        assert list(emitted.inactive_indices().indices(alpha)) == core

    def test_empty_external_space_is_the_identity(self, water_window):
        """Keeping the whole window leaves nothing to fold, so the integrals come back unchanged."""
        case = water_window
        h_eff = algorithms.create(_TYPE, "qdk_swpt2").run(case.reference, case.hamiltonian, case.window_space)

        assert h_eff.get_core_energy() == pytest.approx(case.hamiltonian.get_core_energy(), abs=1e-10)
        one_body, _ = h_eff.get_one_body_integrals()
        window_one_body, _ = case.hamiltonian.get_one_body_integrals()
        np.testing.assert_allclose(one_body, window_one_body, atol=1e-10)
        np.testing.assert_allclose(
            h_eff.get_two_body_integrals()[0], case.hamiltonian.get_two_body_integrals()[0], atol=1e-10
        )

    def test_downfold_dresses_the_two_body_block_and_keeps_four_fold_symmetry(self, water_window):
        """The dressing is nonzero, hermitian, and -- as documented -- not 8-fold symmetric."""
        case = water_window
        h_eff = algorithms.create(_TYPE, "qdk_swpt2").run(case.reference, case.hamiltonian, case.kept_space)

        kept = case.kept_in_window
        n = len(kept)
        window_one_body, _ = case.hamiltonian.get_one_body_integrals()
        window_norb = window_one_body.shape[0]
        g = np.asarray(h_eff.get_two_body_integrals()[0]).reshape(n, n, n, n)
        window_g = np.asarray(case.hamiltonian.get_two_body_integrals()[0])
        assert window_g.size == window_norb**4
        bare = window_g.reshape((window_norb,) * 4)
        bare = bare[np.ix_(kept, kept, kept, kept)]

        assert np.abs(g - bare).max() > 1e-6, "the downfold must actually dress the kept space"
        np.testing.assert_allclose(g, g.transpose(1, 0, 3, 2), atol=1e-10)  # hermiticity
        np.testing.assert_allclose(g, g.transpose(2, 3, 0, 1), atol=1e-10)  # electron exchange
        # The bra swap of a genuine Coulomb integral does not survive the commutator, so
        # consumers reading only the 8-fold-unique elements get a different operator.
        assert np.abs(g - g.transpose(1, 0, 2, 3)).max() > 1e-6

    def test_rejects_non_finite_regularizer(self, water_window):
        """A non-finite sigma must fail loudly rather than silently selecting the bare path."""
        case = water_window
        downfolder = algorithms.create(_TYPE, "qdk_swpt2")
        downfolder.settings().set("regularizer_sigma2", float("nan"))
        with pytest.raises((ValueError, RuntimeError)):
            downfolder.run(case.reference, case.hamiltonian, case.kept_space)

File: cpp/tests/test_effective_hamiltonian.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE.txt in the project root for
// license information.

#include <gtest/gtest.h>

#include <Eigen/Dense>
#include <memory>
#include <optional>
#include <qdk/chemistry/algorithms/effective_hamiltonian.hpp>
#include <qdk/chemistry/data/basis_set.hpp>
#include <qdk/chemistry/data/configuration.hpp>
#include <qdk/chemistry/data/hamiltonian_containers/canonical_four_center.hpp>
#include <qdk/chemistry/data/symmetry/spin_channel_indices.hpp>
#include <qdk/chemistry/data/wavefunction_containers/state_vector.hpp>
#include <string>
#include <vector>

#include "ut_common.hpp"

using namespace qdk::chemistry::algorithms;
using namespace qdk::chemistry::data;

class TestEffectiveHamiltonianConstructor
    : public EffectiveHamiltonianConstructor {
 public:
  std::string name() const override {
    return "_test_effective_hamiltonian_constructor";
  }

  using EffectiveHamiltonianConstructor::_validate_inputs;

 protected:
  std::shared_ptr<Hamiltonian> _run_impl(
      std::shared_ptr<Wavefunction>, std::shared_ptr<Hamiltonian> hamiltonian,
      std::shared_ptr<const SymmetryBlockedIndexSet>) const override {
    return hamiltonian;
  }
};

namespace {

std::shared_ptr<Hamiltonian> make_hamiltonian(
    std::shared_ptr<Orbitals> orbitals) {
  const auto size =
      spin_channel_indices(orbitals->active_indices(), axes::alpha()).size();
  return std::make_shared<Hamiltonian>(
      std::make_unique<CanonicalFourCenterHamiltonianContainer>(
          Eigen::MatrixXd::Identity(size, size),
          Eigen::VectorXd::Zero(size * size * size * size), std::move(orbitals),
          0.0, Eigen::MatrixXd::Zero(0, 0)));
}

std::shared_ptr<Hamiltonian> make_unrestricted_hamiltonian(
    std::shared_ptr<Orbitals> orbitals) {
  const auto size =
      spin_channel_indices(orbitals->active_indices(), axes::alpha()).size();
  const auto num_two_body = size * size * size * size;
  return std::make_shared<Hamiltonian>(
      std::make_unique<CanonicalFourCenterHamiltonianContainer>(
          Eigen::MatrixXd::Identity(size, size),
          Eigen::MatrixXd::Identity(size, size),
          Eigen::VectorXd::Zero(num_two_body),
          Eigen::VectorXd::Zero(num_two_body),
          Eigen::VectorXd::Zero(num_two_body), std::move(orbitals), 0.0,
          Eigen::MatrixXd::Zero(0, 0), Eigen::MatrixXd::Zero(0, 0)));
}

std::shared_ptr<Wavefunction> make_wavefunction(
    std::shared_ptr<Orbitals> orbitals) {
  const auto size =
      spin_channel_indices(orbitals->active_indices(), axes::alpha()).size();
  std::string configuration(size, '0');
  if (!configuration.empty()) configuration[0] = '2';
  return std::make_shared<Wavefunction>(std::make_unique<StateVectorContainer>(
      Configuration::from_spin_half_string(configuration),
      std::move(orbitals)));
}

}  // namespace

TEST(EffectiveHamiltonianConstructorTest, MetaData) {
  TestEffectiveHamiltonianConstructor constructor;
  EXPECT_NO_THROW({ auto settings = constructor.settings(); });
  EXPECT_EQ(constructor.type_name(), "effective_hamiltonian_constructor");
}

TEST(EffectiveHamiltonianConstructorTest, Factory) {
  EXPECT_TRUE(EffectiveHamiltonianConstructorFactory::available().empty());
  EXPECT_THROW(
      EffectiveHamiltonianConstructorFactory::create("nonexistent_constructor"),
      std::runtime_error);
  EXPECT_NO_THROW(EffectiveHamiltonianConstructorFactory::register_instance(
      []() -> EffectiveHamiltonianConstructorFactory::return_type {
        return std::make_unique<TestEffectiveHamiltonianConstructor>();
      }));
  EXPECT_THROW(
      EffectiveHamiltonianConstructorFactory::register_instance(
          []() -> EffectiveHamiltonianConstructorFactory::return_type {
            return std::make_unique<TestEffectiveHamiltonianConstructor>();
          }),
      std::runtime_error);

  auto constructor = EffectiveHamiltonianConstructorFactory::create(
      "_test_effective_hamiltonian_constructor");
  EXPECT_NE(constructor, nullptr);

  EXPECT_FALSE(EffectiveHamiltonianConstructorFactory::unregister_instance(
      "nonexistent_constructor"));
  EXPECT_TRUE(EffectiveHamiltonianConstructorFactory::unregister_instance(
      "_test_effective_hamiltonian_constructor"));
  EXPECT_FALSE(EffectiveHamiltonianConstructorFactory::unregister_instance(
      "_test_effective_hamiltonian_constructor"));
}

TEST(EffectiveHamiltonianConstructorTest, AcceptsNestedOrbitalSpaces) {
  auto base_orbitals = testing::create_test_orbitals(4, 4);
  auto hamiltonian_orbitals =
      testing::with_active_space(base_orbitals, {0, 1, 2, 3}, {});
  auto wavefunction_orbitals =
      testing::with_active_space(base_orbitals, {1, 2, 3}, {0});
  auto hamiltonian = make_hamiltonian(hamiltonian_orbitals);
  auto reference = make_wavefunction(wavefunction_orbitals);
  auto p_indices = testing::restricted_index_set(4, {1, 3});

  TestEffectiveHamiltonianConstructor constructor;
  EXPECT_NO_THROW(
      constructor._validate_inputs(reference, hamiltonian, p_indices));
}

TEST(EffectiveHamiltonianConstructorTest,
     AcceptsNestedUnrestrictedOrbitalSpaces) {
  auto base_orbitals = testing::create_test_orbitals(4, 4, false);
  auto hamiltonian_orbitals =
      testing::with_active_space(base_orbitals, {0, 1, 2, 3}, {});
  auto wavefunction_orbitals =
      testing::with_active_space(base_orbitals, {1, 2, 3}, {0});
  auto hamiltonian = make_unrestricted_hamiltonian(hamiltonian_orbitals);
  auto reference = make_wavefunction(wavefunction_orbitals);
  auto p_indices = testing::unrestricted_index_set(4, {1, 3}, {1, 3});

  TestEffectiveHamiltonianConstructor constructor;
  EXPECT_NO_THROW(
      constructor._validate_inputs(reference, hamiltonian, p_indices));
}

TEST(EffectiveHamiltonianConstructorTest, TreatsPIndicesAsAbsoluteIndices) {
  auto base_orbitals = testing::create_test_orbitals(8, 8);
  auto orbitals = testing::with_active_space(base_orbitals, {0, 2, 4, 6}, {});
  auto hamiltonian = make_hamiltonian(orbitals);
  auto reference = make_wavefunction(orbitals);
  TestEffectiveHamiltonianConstructor constructor;

  EXPECT_NO_THROW(constructor._validate_inputs(
      reference, hamiltonian, testing::restricted_index_set(8, {0, 4})));

  // {1, 3} would name the second and fourth active orbitals if the indices
  // were positions within the active space; as absolute indices they are not
  // in it at all.
  EXPECT_THROW(
      constructor._validate_inputs(reference, hamiltonian,
                                   testing::restricted_index_set(8, {1, 3})),
      std::invalid_argument);
}

TEST(EffectiveHamiltonianConstructorTest, RejectsIncompatibleOrbitalBases) {
  auto hamiltonian_orbitals = testing::create_test_orbitals(4, 4);
  auto hamiltonian = make_hamiltonian(hamiltonian_orbitals);
  TestEffectiveHamiltonianConstructor constructor;

  auto different_size = testing::create_test_orbitals(4, 3);
  EXPECT_THROW(constructor._validate_inputs(
                   make_wavefunction(different_size), hamiltonian,
                   testing::restricted_index_set(3, {0})),
               std::invalid_argument);

  const auto& coefficients = hamiltonian_orbitals->coefficients()->block(
      {axes::alpha(), axes::alpha()});
  auto unrestricted = std::make_shared<Orbitals>(
      coefficients, coefficients, std::nullopt, std::nullopt, std::nullopt,
      hamiltonian_orbitals->get_basis_set(),
      testing::unrestricted_index_set(4, {0, 1, 2, 3}, {0, 1, 2, 3}));
  EXPECT_THROW(constructor._validate_inputs(
                   make_wavefunction(unrestricted), hamiltonian,
                   testing::unrestricted_index_set(4, {0}, {0})),
               std::invalid_argument);
}

TEST(EffectiveHamiltonianConstructorTest, RejectsDifferentAtomicOrbitalBases) {
  auto hamiltonian_orbitals = testing::create_test_orbitals(4, 4);
  auto hamiltonian = make_hamiltonian(hamiltonian_orbitals);

  // Same atom count and same number of atomic orbitals, different exponents.
  auto structure = std::make_shared<Structure>(
      std::vector<Eigen::Vector3d>{{0.0, 0.0, 0.0}, {1.0, 0.0, 0.0}},
      std::vector<Element>{Element::H, Element::H});
  Eigen::VectorXd contraction(1);
  contraction << 1.0;
  Eigen::VectorXd p_exponent(1);
  p_exponent << 2.5;
  Eigen::VectorXd s_exponent(1);
  s_exponent << 3.5;
  std::vector<Shell> shells{Shell(0, OrbitalType::P, p_exponent, contraction),
                            Shell(1, OrbitalType::S, s_exponent, contraction)};
  auto other_basis = std::make_shared<BasisSet>("test", shells, structure);

  const auto& coefficients = hamiltonian_orbitals->coefficients()->block(
      {axes::alpha(), axes::alpha()});
  auto other_orbitals = std::make_shared<Orbitals>(coefficients, std::nullopt,
                                                   std::nullopt, other_basis);

  TestEffectiveHamiltonianConstructor constructor;
  EXPECT_THROW(constructor._validate_inputs(
                   make_wavefunction(other_orbitals), hamiltonian,
                   testing::restricted_index_set(4, {0})),
               std::invalid_argument);
}

TEST(EffectiveHamiltonianConstructorTest, RejectsNonNestedOrbitalSpaces) {
  auto base_orbitals = testing::create_test_orbitals(4, 4);
  auto hamiltonian_orbitals =
      testing::with_active_space(base_orbitals, {0, 1, 2}, {3});
  auto wavefunction_orbitals =
      testing::with_active_space(base_orbitals, {1, 2, 3}, {0});
  auto hamiltonian = make_hamiltonian(hamiltonian_orbitals);
  auto reference = make_wavefunction(wavefunction_orbitals);
  TestEffectiveHamiltonianConstructor constructor;

  EXPECT_THROW(
      constructor._validate_inputs(reference, hamiltonian,
                                   testing::restricted_index_set(4, {1, 2})),
      std::invalid_argument);

  auto valid_wavefunction_orbitals =
      testing::with_active_space(base_orbitals, {1, 2}, {0});
  auto valid_reference = make_wavefunction(valid_wavefunction_orbitals);
  EXPECT_THROW(
      constructor._validate_inputs(valid_reference, hamiltonian,
                                   testing::restricted_index_set(4, {0})),
      std::invalid_argument);
  EXPECT_THROW(
      constructor._validate_inputs(valid_reference, hamiltonian,
                                   testing::restricted_index_set(5, {1})),
      std::invalid_argument);
}

TEST(EffectiveHamiltonianConstructorTest, RejectsNullInputs) {
  auto orbitals = testing::create_test_orbitals(2, 2);
  auto hamiltonian = make_hamiltonian(orbitals);
  auto reference = make_wavefunction(orbitals);
  auto p_indices = testing::restricted_index_set(2, {0});
  TestEffectiveHamiltonianConstructor constructor;

  EXPECT_THROW(constructor._validate_inputs(nullptr, hamiltonian, p_indices),
               std::invalid_argument);
  EXPECT_THROW(constructor._validate_inputs(reference, nullptr, p_indices),
               std::invalid_argument);
  EXPECT_THROW(constructor._validate_inputs(reference, hamiltonian, nullptr),
               std::invalid_argument);
}
